typed-racket-lib/typed-racket/typecheck/tc-envops.rkt

```diff
@@ -29,7 +29,20 @@
   (cond
     [(or (null? ps) (andmap TrueProp? ps)) env]
     [else
-     (define-values (props atoms) (combine-props ps (env-props env)))
+     (define-values (props atoms^) (combine-props ps (env-props env)))
+
+     (define atoms (if atoms^
+                       ;; Fix the order of paths to the same object.
+                       ;; Objects without path elements always come last
+                       (sort atoms^ (lambda (x y)
+                                      (match* (x y)
+                                        [((TypeProp: (Path: pes1 (? identifier? var1)) _)
+                                          (TypeProp: (Path: pes2 (? identifier? var2)) _))
+                                         #:when (equal? var1 var2)
+                                         (not (null? pes1))]
```

Comment:
Shouldn't this use a comparison that's more useful in more cases, like `(< (length pes1 pes2))`, or something like that? In general I worry about a comparison function that isn't an order.

Reply:
my first attempt was indeed using `(< (length pes 1 pes2))`, but I was not so sure it was completely right. 
I will write some tests to make myself confident about the general solution.